lgc/patch/PatchWorkarounds.cpp

```diff
@@ -0,0 +1,141 @@
+  return new PatchWorkarounds();
+}
+
+PatchWorkarounds::PatchWorkarounds() : FunctionPass(ID) {
+}
+
+// =====================================================================================================================
+// Executes this LLVM pass on the specified LLVM function.
+//
+// @param [in/out] function : Function that we will add workarounds.
+bool PatchWorkarounds::runOnFunction(Function &function) {
+  LLVM_DEBUG(dbgs() << "Run the pass Patch-Workarounds\n");
+
+  m_pipelineState = getAnalysis<PipelineStateWrapper>().getPipelineState(function.getParent());
+  m_context = &function.getContext();
+  m_builder = std::make_unique<IRBuilder<>>(*m_context);
+  m_processed.clear();
+
+  m_changed = false;
+
+  // At present we only apply a single workaround in this pass. Makes sense to check here an early exit the pass if not
+  // required
+  if (!m_pipelineState->getOptions().disableImageResourceCheck &&
+      m_pipelineState->getTargetInfo().getGpuWorkarounds().gfx10.waFixBadImageDescriptor)
+    visit(function);
+
+  return m_changed;
+}
+
+// =====================================================================================================================
+// Visits "call" instruction.
+//
+// @param callInst : The instruction
+void PatchWorkarounds::visitCallInst(CallInst &callInst) {
+  Function *const calledFunc = callInst.getCalledFunction();
+  if (!calledFunc)
+    return;
+
+  auto name = calledFunc->getName();
+
+  const bool isImage = name.startswith("llvm.amdgcn.image.");
+
+  // A buffer descriptor may be incorrectly given when it should be an image descriptor, we need to fix it to valid
+  // buffer type (0) to make hardware happily ignore it. This is to check and fix against buggy applications which
+  // declares a image descriptor in shader but provide a buffer descriptor in driver. Note this only applies to gfx10.1.
+  if (isImage) {
+    // Look for 8 dword resource descriptor argument
+    for (Value *arg : callInst.args()) {
+      if (auto vecTy = dyn_cast<FixedVectorType>(arg->getType())) {
+        if (vecTy->getNumElements() == 8 && vecTy->getElementType()->isIntegerTy(32)) {
+          if (isa<UndefValue>(arg))
+            // We don't need to worry if the value is actually undef
+            break;
+
+          if (m_processed.find(arg) != m_processed.end())
+            // Already processed this one
+            break;
+
+          m_processed.insert(arg);
+
+          Instruction *rsrcInstr = cast<Instruction>(arg);
+
+          m_builder->SetInsertPoint(rsrcInstr->getNextNode());
+          m_builder->SetCurrentDebugLocation(rsrcInstr->getNextNode()->getDebugLoc());
+
+          // Create a dummy Value to temporarily replace the descriptor- we'll replace all uses of this later
+          Value *tempVal = m_builder->CreateExtractElement(rsrcInstr, 3);
+          Value *oldUse = m_builder->CreateVectorSplat(vecTy->getElementCount(), tempVal);
+          rsrcInstr->replaceAllUsesWith(oldUse);
```

Comment:
I think you can simplify the code with ` newInst = rsrcInst->clone(); ` You can set a new name to the `newInst` and insert into the basicblock, then you can reference the `newInst` when inserting workaround instructions and at the very last ` rsrcInst->replaceAllUsesWith(newArg); `

Reply:
Yes - good idea. That's much better.

Reply:
Done.